jellyfin/jellyfin-androidtv
# Add Quality Selection and Audio Output Controls to Playback Overlay

**Changes**


1. Added quality selection (max bitrate) menu to playback overlay.
2. Added audio output (downmixing/transcoding) selection menu to playback overlay.
3. Added PlaybackController.refreshSteam() function to restart a stream when transcoding options are changed.
4. Fixed Mbit/s to Kbit/s conversion factor (100 -> 1000). 
5. Fixed unusual Mbit/s increment (21Mbit/s -> 20Mbit/s).

This PR builds off of https://github.com/jellyfin/jellyfin-androidtv/pull/1302 using my _very_ limited understanding of Kotlin, implementing `VideoQualityController` / `SelectQualityAction` and `AudioOutputController` / `SelectAudioOutputAction`.

As a result of this change, it was necessary to implement the function `PlaybackController.refreshSteam()` in order to restart the stream when the setting is changed, applying the new setting.

During this change, it was also noted that the Mbit/s to Kbit/s conversion factor was incorrect (x100), and an unusual quality increment had been added to the settings menu (21 Mbit/s). These have been fixed.

**Issues**
May satisfy these vague enhancements:
- https://github.com/jellyfin/jellyfin-androidtv/issues/1388
- https://github.com/jellyfin/jellyfin-androidtv/issues/455
- https://github.com/jellyfin/jellyfin-androidtv/issues/43


**Testing**
- Try changing quality / audio output settings through the playback menu. You should be able to verify that the media is being transcoded / downmixed through the logs / dashboard.
- Try playing a different video - the settings should persist.
- Open the settings -> playback menu through the main menu - the settings configured through the overlay should persist there.
- Change the quality / audio output settings in the settings -> playback menu through the main menu, then open both menus in the playback overlay - the settings should persist and you should be able to verify that the media is being transcoded / downmixed through the logs / dashboard.
